pkg/clustermesh/kvstoremesh/remote_cluster.go

```diff
@@ -130,8 +133,74 @@ func (rc *remoteCluster) Stop() {
 }
 
 func (rc *remoteCluster) Remove() {
-	// Cluster specific keys are not explicitly removed, but they will be
-	// disappear once the associated lease expires.
+	const retries = 5
+
+	fn, done := rc.drainJobFn(retries)
+	rc.jobGroup.Add(job.OneShot(
+		fmt.Sprintf("drain-cluster-%s", rc.name), fn,
+		job.WithRetry(retries, &job.ExponentialBackoff{Min: 2 * time.Second, Max: 60 * time.Second}),
+	))
+
+	<-done
+}
+
+// drainJobFn returns a job function responsible for draining the cached data
+// from the local kvstore.
+func (rc *remoteCluster) drainJobFn(retries uint) (job.OneShotFunc, <-chan struct{}) {
+	var retry uint
+	done := make(chan struct{})
+	keys := []string{
+		path.Join(kvstore.ClusterConfigPrefix, rc.name),
+	}
+	prefixes := []string{
+		path.Join(kvstore.SyncedPrefix, rc.name),
```

Comment:
If I remember correctly, order in which we remove here is probably very important. 
I would love to see some reasoning of what happens to running cilium-agents and bootstrapping cilium-agents. 

For example, if we remove `ClusterConfigPrefix` first, then new clustermesh clients won't know that they should use syncCanaries, which could result in reading a partial state - which probably is fine, but it makes me wonder also about other cases.

Reply:
IMO deleting the cluster config first is the right thing to do here, because that effectively prevents bootstrapping agents from connecting (because they expect the config to be present), which is the expected behavior considering that we are deleting the cached data for that cluster in any case.

The deletion order of the other items doesn't matter once the agents are already connected, and they'll receive the deletion events for the various keys. Regardless, it is expected that agents will disconnect from the same cluster as well, hence they would drain all the keys in any case.

Another possibility would be that the bootstrapping agent already retrieved the cluster config, but hasn't started watching yet. In this case, deleting the synced prefix first prevents again from connecting, which looks correct to me, although that's arguably a rare condition.